### ISSUE TITLE
Inconsistent errors with SparseArray construction (new in 1.2)

### ISSUE THREAD

**Issue by sbromberger** [JuliaLang/julia#32548]
In 1.1:
```
julia> a = spzeros(Float64, UInt8, 255, 255)
255×255 SparseMatrixCSC{Float64,UInt8} with 0 stored entries

julia> a = spzeros(Float64, UInt8, 0xff, 0xff)
255×255 SparseMatrixCSC{Float64,UInt8} with 0 stored entries
```

But in 1.2, there apparently was a change to ensure that the number of nonzero entries could fit:
```
julia> a = spzeros(Float64, UInt8, 255, 255)
ERROR: ArgumentError: maximal nnz+1 (m*n+1 = 65026) does not fit in Ti = UInt8
```

However, the check is not applied consistently:
```
julia> a = spzeros(Float64, UInt8, 0xff, 0xff)
255×255 SparseMatrixCSC{Float64,UInt8} with 0 stored entries
```

**Comment by StefanKarpinski** [comment]
Note that this might be a 1.2 regression so we should take a look at it on triage.

**Comment by sbromberger** [comment]
(Just a guess that the error checks by multiplying the dimensions together to get the maximum possible number of nnzs, and `0xff * 0xff` overflows to `0x01` which, when increased by 1, is still less than `typemax(UInt8)`.)

**Comment by KristofferC** [comment]
This was reverted for 1.2 because it was breaking (https://github.com/JuliaLang/julia/pull/31667). The PR introducing this was merged 2 weeks ago https://github.com/JuliaLang/julia/pull/31724 and has not been backported so I am pretty sure this only exists on master.

**Comment by StefanKarpinski** [comment]
> This was reverted for 1.2

That's a relief. Sorry for the false alarm.

**Comment by JeffBezanson** [comment]
I believe this and #32539 were caused by the same set of changes.

**Comment by sbromberger** [comment]
Thank you very much for the quick analysis.

**Comment by JeffBezanson** [comment]
Fixed by #31724.

**Comment by StefanKarpinski** [comment]
Seems to be fixed.
